File: LeetCode/竞赛/周赛/周赛-243/5774. 使用服务器处理任务.py
import heapq
class Solution:
    def assignTasks(self, servers: List[int], tasks: List[int]) -> List[int]:
        pq = []
        m = len(tasks)
        n = len(servers)
        ans = [None for _ in range(m)]
        for i in range(n):
            heapq.heappush(pq, [servers[i], i])
        pqing = []
        
        time = 0
        idx = 0
        while True:
            if idx == m:
                break
            if time < idx:
                time = idx
            if not pq:
                p = heapq.heappop(pqing)
                time = p[0]
                heapq.heappush(pq, p[1:])
            while pqing:
                p = heapq.heappop(pqing)
                if p[0] > time:
                    heapq.heappush(pqing, p)
                    break
                heapq.heappush(pq, p[1:])

            p = heapq.heappop(pq)
            heapq.heappush(pqing, [time+tasks[idx], p[0], p[1]])
            ans[idx] = p[1]
            idx += 1
        return ans

File: LeetCode/竞赛/周赛/周赛-243/5774. 使用服务器处理任务.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<int> assignTasks(vector<int>& servers, vector<int>& tasks) {
        vector<int> ans;
        int a, b, c;
        vector<int> p;
        priority_queue<vector<int>, vector<vector<int>>, greater<int> > pq;
        for (int i=0; i < servers.size(); i++) {
            vector<int> v{0, servers[i], i};
            pq.push(v);
        }
        int idx = 0;
        for (int time=1; ; time++) {
            if (idx == tasks.size()) break;
            p = pq.top();
            pq.pop();
            if (p[0] > time) {
                time = p[0];
            }
            ans.push_back(p[2]);
            p[0] += tasks[idx];
            pq.push(p);
            idx++;
        }
        return ans;
    }
};